packages/curve-ui-kit/src/shared/ui/DataTable.tsx

```diff
@@ -0,0 +1,158 @@
+import Typography from '@mui/material/Typography'
+import TableHead from '@mui/material/TableHead'
+import TableCell from '@mui/material/TableCell'
+import TableBody from '@mui/material/TableBody'
+import Table from '@mui/material/Table'
+import TableRow from '@mui/material/TableRow'
+import { SizesAndSpaces } from '../../themes/design/1_sizes_spaces'
+import { useIntersectionObserver } from 'ui'
+import {
+  Cell,
+  Column,
+  ColumnDef,
+  flexRender,
+  getCoreRowModel,
+  getSortedRowModel,
+  Header,
+  Row,
+  SortingState,
+  useReactTable,
+} from '@tanstack/react-table'
+import { useSortFromQueryString } from '../../hooks/useSortFromQueryString'
+import { ArrowDownIcon } from '../icons/ArrowDownIcon'
+
+const { Sizing, Spacing, MinWidth } = SizesAndSpaces
+
+const getAlignment = <T extends unknown>({ columnDef }: Column<T>) =>
+  columnDef.meta?.type == 'numeric' ? 'right' : 'left'
+
+const DataCell = <T extends unknown>({ cell }: { cell: Cell<T, unknown> }) => {
+  const column = cell.column
+  return (
+    <Typography
+      variant={column.columnDef.meta?.variant ?? 'tableCellMBold'}
+      color="text.primary"
+      component="td"
+      sx={{
+        textAlign: getAlignment(column),
+        paddingInline: Spacing.sm,
+        paddingBlock: Spacing.md,
+        ...getExtraColumnPadding(column),
+      }}
+      data-testid={`data-table-cell-${column.id}`}
+    >
+      {flexRender(column.columnDef.cell, cell.getContext())}
+    </Typography>
+  )
+}
+
+const DataRow = <T extends unknown>({ row }: { row: Row<T> }) => {
+  const ref = useRef<HTMLTableRowElement>(null)
+  const entry = useIntersectionObserver(ref, { freezeOnceVisible: true })
+  return (
+    <TableRow
+      sx={{
+        marginBlock: 0,
+        height: Sizing['3xl'],
+        borderBottom: '1px solid',
+        borderColor: (t) => t.design.Layer[1].Outline,
+      }}
+      ref={ref}
+      data-testid={`data-table-row-${row.id}`}
+    >
+      {/* render cells when visible vertically, so content is lazy loaded */}
+      {entry?.isIntersecting && row.getVisibleCells().map((cell) => <DataCell key={cell.id} cell={cell} />)}
+    </TableRow>
+  )
+}
+
+/**
+ * In the figma design, the first and last columns seem to be aligned to the table title.
+ * However, the normal padding causes them to be misaligned.
+ */
+const getExtraColumnPadding = <T extends any>(column: Column<T>) => ({
+  ...(column.getIsFirstColumn() && { paddingInlineStart: Spacing.md }),
+  ...(column.getIsLastColumn() && { paddingInlineEnd: Spacing.md }),
+})
+
+const HeaderCell = <T extends unknown>({ header }: { header: Header<T, unknown> }) => {
+  const { column } = header
+  const sort = column.getIsSorted()
+  return (
+    <TableCell
+      sx={{
+        textAlign: getAlignment(column),
+        alignContent: 'end',
+        padding: Spacing.sm,
+        paddingBlockStart: 0,
+        ...getExtraColumnPadding(column),
+        ...(column.columnDef.enableSorting && {
```

Comment:
This doesn't work, because if `enableSorting` isn't set the default behavior is indeed to be `true`, but the actual value of the property is `undefined` thus this is false. Use `column.getCanSort()` instead.

Reply:
yeah it should have been enableSorting != false but getCanSort is better indeed